### PR TITLE
 small normalization improvement

### PR DESCRIPTION
**Pull request by lcnr**
r? @compiler-errors

### CODE REVIEW

File: compiler/rustc_type_ir/src/flags.rs [hunk 1]
```diff
@@ -81,7 +81,7 @@ bitflags::bitflags! {
         /// Does this have `Alias` or `ConstKind::Unevaluated`?
         ///
         /// Rephrased, could this term be normalized further?
-        const HAS_ALIASES              = TypeFlags::HAS_TY_PROJECTION.bits()
+        const HAS_ALIAS                   = TypeFlags::HAS_TY_PROJECTION.bits()
```
Comment (fmease) [inline compiler/rustc_type_ir/src/flags.rs:84]: 👍

File: compiler/rustc_type_ir/src/visit.rs [hunk 2]
```diff
@@ -230,7 +230,7 @@ pub trait TypeVisitableExt<I: Interner>: TypeVisitable<I> {
     }
 
     fn has_aliases(&self) -> bool {
```
Comment (fmease) [inline compiler/rustc_type_ir/src/visit.rs:232]:
```suggestion
     fn has_alias(&self) -> bool {
```

too (my bad)

Comment (lcnr) [reply compiler/rustc_type_ir/src/visit.rs:232]: interstingly, we do use the plural for the functions, e.g. we have `has_inherent_projections` and `has_opaque_types` even though the type flag is `HAS_TY_OPAQUE`

Comment (compiler-errors) [reply compiler/rustc_type_ir/src/visit.rs:232]: let's leave this one

File: compiler/rustc_trait_selection/src/traits/normalize.rs [hunk 3]
```diff
@@ -109,16 +109,13 @@ pub(super) fn needs_normalization<'tcx, T: TypeVisitable<TyCtxt<'tcx>>>(
...
-        | ty::TypeFlags::HAS_TY_INHERENT
-        | ty::TypeFlags::HAS_CT_PROJECTION;
+    let mut flags = ty::TypeFlags::HAS_ALIAS;
 
+    // Opaques are treated as rigid with `Reveal::UserFacing`,
+    // so we can ignore those.
     match reveal {
-        Reveal::UserFacing => {}
-        Reveal::All => flags |= ty::TypeFlags::HAS_TY_OPAQUE,
+        Reveal::UserFacing => flags ^= ty::TypeFlags::HAS_TY_OPAQUE,
```
Comment (compiler-errors) [inline compiler/rustc_trait_selection/src/traits/normalize.rs:117]: is it possible to & with the complement here? xor feels confusing to reason about

Comment (lcnr) [reply compiler/rustc_trait_selection/src/traits/normalize.rs]: there's `fn remove` :+1:

**Review by compiler-errors — APPROVED** [review]

File: compiler/rustc_trait_selection/src/traits/normalize.rs [hunk 4]
```diff
@@ -109,16 +109,13 @@ pub(super) fn needs_normalization<'tcx, T: TypeVisitable<TyCtxt<'tcx>>>(
...
-        | ty::TypeFlags::HAS_CT_PROJECTION;
+    let mut flags = ty::TypeFlags::HAS_ALIAS;
 
+    // Opaques are treated as rigid with `Reveal::UserFacing`,
+    // so we can ignore those.
     match reveal {
-        Reveal::UserFacing => {}
-        Reveal::All => flags |= ty::TypeFlags::HAS_TY_OPAQUE,
+        Reveal::UserFacing => flags.remove(ty::TypeFlags::HAS_TY_OPAQUE),
+        Reveal::All => {}
```
Comment (klensy) [inline compiler/rustc_trait_selection/src/traits/normalize.rs:118]: Logic here now inverted, is it ok?

Comment (klensy) [reply compiler/rustc_trait_selection/src/traits/normalize.rs:118]: Ahh, `HAS_TY_OPAQUE` wasn't in `flags` before.

**Pull request merged**